components/helm-broker/internal/controller/addonsconfiguration_controller.go

```diff
@@ -0,0 +1,500 @@
+			addon.Charts = completeBundle.Charts
+
+			adds = append(adds, addon)
+		}
+	}
+
+	return adds, nil
+}
+
+func (r *ReconcileAddonsConfiguration) saveBundle(namespace internal.Namespace, repositories *addons.RepositoryCollection) error {
+	for _, addon := range repositories.ReadyAddons() {
+		if len(addon.Bundle.Docs) == 1 {
+			r.log.Infof("- ensure DocsTopic for bundle %s in namespace %s", addon.Bundle.ID, namespace)
+			if err := r.docsTopicProvider.EnsureDocsTopic(addon.Bundle, string(namespace)); err != nil {
+				return exerr.Wrapf(err, "While ensuring DocsTopic for bundle %s/%s: %v", addon.Bundle.ID, namespace, err)
+			}
+		}
+		exist, err := r.bundleStorage.Upsert(namespace, addon.Bundle)
+		if err != nil {
+			addon.RegisteringError(err)
+			r.log.Errorf("cannot upsert bundle %v:%v into storage", addon.Bundle.Name, addon.Bundle.Version)
+			continue
+		}
+		if exist {
+			r.log.Infof("bundle %v:%v already existed in storage, bundle was replaced", addon.Bundle.Name, addon.Bundle.Version)
+		}
+		err = r.saveCharts(namespace, addon.Charts)
+		if err != nil {
+			addon.RegisteringError(err)
+			r.log.Errorf("cannot upsert charts of %v:%v bunlde", addon.Bundle.Name, addon.Bundle.Version)
+			continue
+		}
+
+		r.syncBroker = true
+	}
+	return nil
+}
+
+func (r *ReconcileAddonsConfiguration) saveCharts(namespace internal.Namespace, charts []*chart.Chart) error {
+	for _, bundleChart := range charts {
+		exist, err := r.chartStorage.Upsert(namespace, bundleChart)
+		if err != nil {
+			r.log.Errorf("cannot upsert %s chart: %s", bundleChart.Metadata.Name, err)
+			return err
+		}
+		if exist {
+			r.log.Infof("chart %s already existed in storage, chart was replaced", bundleChart.Metadata.Name)
+		}
+	}
+	return nil
+}
+
+func (r *ReconcileAddonsConfiguration) statusSnapshot(addon *addonsv1alpha1.AddonsConfiguration, repositories *addons.RepositoryCollection) {
+	addon.Status.Repositories = nil
+
+	for _, repo := range repositories.Repositories {
+		addonsRepository := repo.Repository
+		addonsRepository.Addons = []addonsv1alpha1.Addon{}
+		for _, addon := range repo.Addons {
+			addonsRepository.Addons = append(addonsRepository.Addons, addon.Addon)
+		}
+		addon.Status.Repositories = append(addon.Status.Repositories, addonsRepository)
+	}
+
+	if repositories.IsRepositoriesIDConflict() {
+		addon.Status.Phase = addonsv1alpha1.AddonsConfigurationFailed
+	} else {
+		addon.Status.Phase = addonsv1alpha1.AddonsConfigurationReady
+	}
+}
+
+func (r *ReconcileAddonsConfiguration) updateAddonStatus(addon *addonsv1alpha1.AddonsConfiguration) error {
+	instance := &addonsv1alpha1.AddonsConfiguration{}
+	err := r.Get(context.TODO(), types.NamespacedName{Name: addon.Name, Namespace: addon.Namespace}, instance)
+	if err != nil {
+		return exerr.Wrap(err, "while getting AddonsConfiguration")
+	}
+
+	instance.Status.ObservedGeneration = instance.Generation
+	instance.Status.LastProcessedTime = &v1.Time{Time: time.Now()}
+
```

Comment:
you pass `addon` as a parameter with fulfilled status, and before update it, you fetch `addon` from cache(?) and change only two fields. Do not you lose all the information about status?